cpp/src/arrow/compute/exec/expression.cc

```diff
@@ -394,14 +394,13 @@ Result<Expression> BindImpl(Expression expr, const TypeOrSchema& in,
   if (expr.literal()) return expr;
 
   if (auto ref = expr.field_ref()) {
-    if (ref->IsNested()) {
-      return Status::NotImplemented("nested field references");
-    }
-
     ARROW_ASSIGN_OR_RAISE(auto path, ref->FindOne(in));
 
     auto bound = *expr.parameter();
-    bound.index = path[0];
+    bound.indices.resize(path.indices().size());
+    for (size_t i = 0; i < path.indices().size(); ++i) {
+      bound.indices[i] = path.indices()[i];
+    }
```

Comment:
Perhaps a `std::copy(std::begin(path.indices()), std::end(path.indices()), bound.indices)` would work?

Reply:
It seems SmallVector doesn't implement iterator_traits fully (I also had issues with trying to use insert() that I should try and debug)